src/protocols/bdx/BdxTransferSession.h

```diff
@@ -0,0 +1,361 @@
+/**
+ *    @file
+ *      This file defines a TransferSession state machine that contains the main logic governing a Bulk Data Transfer session. It
+ *      provides APIs for starting a transfer or preparing to receive a transfer request, providing input to be processed, and
+ *      accessing output data (including messages to be sent, message data received by the TransferSession, or state information).
+ */
+
+#pragma once
+
+#include <core/CHIPError.h>
+#include <protocols/bdx/BdxMessages.h>
+#include <system/SystemPacketBuffer.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+namespace BDX {
+
+enum TransferRole : uint8_t
+{
+    kRole_Receiver = 0,
+    kRole_Sender   = 1,
+};
+
+class DLL_EXPORT TransferSession
+{
+public:
+    enum OutputEventFlags : uint16_t
+    {
+        kOutput_None = 0,
+
+        kOutput_MsgToSend      = (1U),
```

Comment:
There's one case I can think of where we might have multiple output events, and that's if an internal error occurs that results in a StatusReport being sent. You would have kOutput_MsgToSend and kOutput_InternalError active simultaneously.

Although, I could create a kStatusReport_ToSend which would include both the message data and the status code, and then enforce having only one output at a time. What do you think?

Reply:
Well, you could instead just internally keep track of the fact that there is sequencing between MsgToSend and InternalError needing to occur one after the other, and always expect the caller to loop-process on PollEvent.

Otherwise, with the mask, you are having a combined weird view of the data, where half of it is in a unioned struct, and the other half is in a message payload. This doesn't scale to a future change where the state machine may need to emit 2 messages in a row, or similarly, use two separate event bodies that are not orthogonal in the output event object.

Reply:
Discussed offline - will change to flat enum